csharp/ql/src/experimental/CWE-095/razor_parse.ql

```diff
@@ -0,0 +1,71 @@
+/**
+ * @name Server-Side Template Injection in RazorEngine
+ * @description User-controlled data may be evaluated, leading to arbitrary code execution.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id csharp/razor-injection
+ * @tags security
+ */
+
+import csharp
+import semmle.code.csharp.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/*
+ * The offending method is Parse from RazorEngine.Razor.
+ */
+
+class RazorEngineClass extends Class {
+  RazorEngineClass() { this.hasQualifiedName("RazorEngine.Razor") }
+
+  Method getParseMethod() {
+    result.getDeclaringType() = this and
+    result.hasName("Parse")
+  }
+}
+
+/*
+ * We are only interested in ASP.NET MVC Controller classes
+ */
+
+class ControllerMVC extends Class {
+  ControllerMVC() { this.hasQualifiedName("System.Web.Mvc", "Controller") }
```

Comment:
New ASP.NET MVC (version 6) has a different namespace. It's possible that this query can cover new MVC if you just include the new namespace (or this.hasQualifiedName("Microsoft.AspNetCore.Mvc", "Controller")

Also, there could be custom controllers derived from ControllerBase. Consider defining namespace separately from the controller (see queries\semmlecode-csharp-queries\semmle\code\csharp\frameworks\microsoft\AspNetCore.qll):
`class MicrosoftAspNetCoreMvcControllerBaseClass extends Class {                                                                                                                   MicrosoftAspNetCoreMvcControllerBaseClass() {                                                                                                                                     getNamespace() = any(MicrosoftAspNetCoreMvcNamespace h) and                                                                                                   (hasName("Controller") or hasName("ControllerBase"))                                                                                                                          }                                                                                                                                                                 }  `

Reply:
Thanks for pointing this out. I've added the support for both the new namespace and ControllerBase classes.